File: bafe/src/ops.c
/* bafe/ops.c - op registry and shape inference
 *
 * We register all ops at startup. The registry is a static array; lookups
 * are by linear scan (we have ~20 ops, no need for a hash table).
 */
#include "bafe/ops.h"
#include <string.h>
#include <stddef.h>

/* ------------------------------------------------------------------ */
/* Attrs                                                              */
/* ------------------------------------------------------------------ */

bafe_op_attrs bafe_op_attrs_default(void) {
    bafe_op_attrs a;
    memset(&a, 0, sizeof(a));
    return a;
}

/* ------------------------------------------------------------------ */
/* Shape inference functions                                          */
/* ------------------------------------------------------------------ */

static bafe_shape _matmul(const bafe_shape *in, int n, const bafe_op_attrs *a) {
    (void)a; (void)n;  /* n == 2 */
    const bafe_shape *A = &in[0];
    const bafe_shape *B = &in[1];
    if (A->rank < 2 || B->rank < 2) return bafe_shape_scalar();
    int32_t M = A->dims[A->rank - 2];
    int32_t K1 = A->dims[A->rank - 1];
    int32_t K2 = B->dims[B->rank - 2];
    int32_t N = B->dims[B->rank - 1];
    (void)K1; (void)K2;  /* mismatch checked at validation, not here */
    if (A->rank == 2 && B->rank == 2) return bafe_shape_make_2(M, N);
    /* batched: broadcast leading dims */
    int32_t ba = A->rank - 2;
    int32_t bb = B->rank - 2;
    int32_t nbatch = ba > bb ? ba : bb;
    bafe_shape batched_a, batched_b;
    batched_a.rank = ba; batched_b.rank = bb;
    for (int32_t i = 0; i < ba; i++) batched_a.dims[i] = A->dims[i];
    for (int32_t i = 0; i < bb; i++) batched_b.dims[i] = B->dims[i];
    bafe_shape batch = bafe_shape_broadcast(&batched_a, &batched_b);
    if (nbatch + 2 > BAFE_MAX_RANK) return bafe_shape_make_2(M, N);
    bafe_shape out;
    out.rank = nbatch + 2;
    for (int32_t i = 0; i < nbatch; i++) out.dims[i] = batch.dims[i];
    out.dims[nbatch] = M;
    out.dims[nbatch + 1] = N;
    return out;
}

static bafe_shape _binary_broadcast(const bafe_shape *in, int n, const bafe_op_attrs *a) {
    (void)a; (void)n;
    return bafe_shape_broadcast(&in[0], &in[1]);
}

static bafe_shape _scale(const bafe_shape *in, int n, const bafe_op_attrs *a) {
    (void)a; (void)n;
    /* scale(tensor, scalar) -> tensor */
    return in[0];
}

static bafe_shape _bias_add(const bafe_shape *in, int n, const bafe_op_attrs *a) {
    (void)a; (void)n;
    /* bias_add([M,N], [N]) -> [M,N] */
    return in[0];
}

static bafe_shape _unary(const bafe_shape *in, int n, const bafe_op_attrs *a) {
    (void)a; (void)n;
    return in[0];
}

static bafe_shape _transpose(const bafe_shape *in, int n, const bafe_op_attrs *a) {
    (void)n;
    return bafe_shape_transpose(&in[0], a->perm);
}

static bafe_shape _reduce(const bafe_shape *in, int n, const bafe_op_attrs *a) {
    (void)n;
    return bafe_shape_reduce(&in[0], a->axes, a->n_axes, a->keepdims);
}

static bafe_shape _reshape(const bafe_shape *in, int n, const bafe_op_attrs *a) {
    (void)n;
    /* resolve -1 in target shape */
    int32_t target[BAFE_MAX_RANK];
    int32_t ntarget = a->n_shape;
    size_t in_numel = bafe_shape_numel(&in[0]);
    for (int32_t i = 0; i < ntarget; i++) target[i] = a->shape[i];
    int32_t neg = -1;
    int32_t known = 1;
    for (int32_t i = 0; i < ntarget; i++) {
        if (target[i] == -1) {
            if (neg != -1) return bafe_shape_scalar(); /* multiple -1 invalid */
            neg = i;
        } else {
            known *= target[i];
        }
    }
    if (neg != -1) {
        if (known == 0 || in_numel % (size_t)known != 0) return bafe_shape_scalar();
        target[neg] = (int32_t)(in_numel / (size_t)known);
    }
    return bafe_shape_make(ntarget, target);
}

static bafe_shape _broadcast_to(const bafe_shape *in, int n, const bafe_op_attrs *a) {
    (void)n; (void)in;
    return bafe_shape_make(a->n_shape, a->shape);
}

static bafe_shape _fused_matmul_bias(const bafe_shape *in, int n, const bafe_op_attrs *a) {
    (void)n; (void)a;
    /* fused_matmul_bias(A, B, bias) -> matmul(A, B) */
    bafe_shape mm_in[2] = {in[0], in[1]};
    return _matmul(mm_in, 2, a);
}

/* ------------------------------------------------------------------ */
/* Registry                                                           */
/* ------------------------------------------------------------------ */

static const bafe_op _OPS[] = {
    {"input",   0, false, "bafe_input",   _unary},  /* shape comes from caller */
    {"constant",0, false, "bafe_constant",_unary},
    {"matmul",  2, false, "bafe_matmul",  _matmul},
    {"add",     2, false, "bafe_add",     _binary_broadcast},
    {"mul",     2, false, "bafe_mul",     _binary_broadcast},
    {"sub",     2, false, "bafe_sub",     _binary_broadcast},
    {"scale",   2, false, "bafe_scale",   _scale},
    {"bias_add",2, false, "bafe_bias_add",_bias_add},
    {"relu",    1, false, "bafe_relu",    _unary},
    {"sigmoid", 1, false, "bafe_sigmoid", _unary},
    {"tanh",    1, false, "bafe_tanh",    _unary},
    {"neg",     1, false, "bafe_neg",     _unary},
    {"layout_transform", 1, false, "bafe_layout_transform", _unary},
    {"transpose",1,false, "bafe_transpose",_transpose},
    {"reduce_sum",1,false,"bafe_reduce_sum",_reduce},
    {"reduce_max",1,false,"bafe_reduce_max",_reduce},
    {"reshape", 1, false, "bafe_reshape", _reshape},
    {"broadcast_to",1,false,"bafe_broadcast_to",_broadcast_to},
    /* fused ops */
    {"fused_matmul_relu",      2, true, "bafe_fused_matmul_relu",      _matmul},
    {"fused_matmul_bias",      3, true, "bafe_fused_matmul_bias",      _fused_matmul_bias},
    {"fused_matmul_bias_relu", 3, true, "bafe_fused_matmul_bias_relu", _fused_matmul_bias},
    {"fused_bias_relu",        2, true, "bafe_fused_bias_relu",        _bias_add},
};

static const int _N_OPS = (int)(sizeof(_OPS) / sizeof(_OPS[0]));

const bafe_op *bafe_op_get(const char *name) {
    for (int i = 0; i < _N_OPS; i++) {
        if (strcmp(_OPS[i].name, name) == 0) return &_OPS[i];
    }
    return NULL;
}

int bafe_op_count(void) { return _N_OPS; }

const bafe_op *bafe_op_at(int i) {
    if (i < 0 || i >= _N_OPS) return NULL;
    return &_OPS[i];
}

bool bafe_op_is_fused(const char *op_name) {
    if (!op_name) return false;
    return strncmp(op_name, "fused_", 6) == 0;
}

bool bafe_op_is_layout_transform(const char *op_name) {
    if (!op_name) return false;
    return strcmp(op_name, "layout_transform") == 0;
}

void bafe_op_registry_init(void) {
    /* nothing to do; the registry is statically initialized */
}

File: bafe/ir/ops.py
"""Op definitions for BAFE IR.

Each op has:
  - a name (str, used in IR and C codegen)
  - an arity (number of tensor inputs)
  - extra attributes (axes, scale factor, etc.) encoded as a frozen dict
  - a shape-inference function

Ops are pure value objects: two Op instances with the same name and attrs
are equal and hash-equal. This matters for the e-graph, which hashes
ENodes (op + child eclass ids).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Mapping, Tuple, Callable, Any

from bafe.ir.types import Shape, Dtype, broadcast_shapes, reduce_shape


# ---------------------------------------------------------------------------
# Op registry
# ---------------------------------------------------------------------------

ShapeFn = Callable[[Tuple[Shape, ...], Mapping[str, Any]], Shape]


@dataclass(frozen=True)
class Op:
    """An op definition (template), not an op instance."""
    name: str
    arity: int
    attrs_schema: Tuple[str, ...] = field(default_factory=tuple)
    shape_fn: ShapeFn = None
    has_fusion_form: bool = False
    c_name: str = ""

    def __post_init__(self):
        if not self.c_name:
            object.__setattr__(self, "c_name", f"bafe_{self.name}")

    def infer_shape(self, shapes: Tuple[Shape, ...], attrs: Mapping[str, Any]) -> Shape:
        if self.shape_fn is None:
            raise NotImplementedError(f"op {self.name} has no shape_fn")
        return self.shape_fn(shapes, attrs)


_REGISTRY: dict[str, Op] = {}


def register(
    name: str,
    arity: int,
    *,
    attrs_schema: Tuple[str, ...] = (),
    has_fusion_form: bool = False,
    c_name: str = "",
) -> Callable[[ShapeFn], ShapeFn]:
    def deco(fn: ShapeFn) -> ShapeFn:
        op = Op(
            name=name,
            arity=arity,
            attrs_schema=tuple(attrs_schema),
            shape_fn=fn,
            has_fusion_form=has_fusion_form,
            c_name=c_name or f"bafe_{name}",
        )
        if name in _REGISTRY:
            raise ValueError(f"op {name!r} already registered")
        _REGISTRY[name] = op
        return fn
    return deco


def get_op(name: str) -> Op:
    if name not in _REGISTRY:
        raise KeyError(f"unknown op {name!r}; registered: {sorted(_REGISTRY)}")
    return _REGISTRY[name]


def all_ops() -> Tuple[Op, ...]:
    return tuple(_REGISTRY.values())


def is_fused(op_name: str) -> bool:
    return op_name.startswith("fused_")


# ---------------------------------------------------------------------------
# Shape inference functions
# ---------------------------------------------------------------------------

@register("matmul", 2, c_name="bafe_matmul")
def _shape_matmul(shapes: Tuple[Shape, ...], attrs: Mapping) -> Shape:
    a, b = shapes
    if a.rank < 2 or b.rank < 2:
        raise ValueError(f"matmul needs rank>=2 inputs, got {a}, {b}")
    m, k1 = a.dims[-2], a.dims[-1]
    k2, n = b.dims[-2], b.dims[-1]
    if k1 != k2:
        raise ValueError(f"matmul K mismatch: {a} x {b}")
    if a.rank == 2 and b.rank == 2:
        return Shape.of(m, n)
    ba = a.dims[:-2]
    bb = b.dims[:-2]
    bbroad = broadcast_shapes(Shape(dims=ba), Shape(dims=bb))
    return Shape(dims=bbroad.dims + (m, n))


def _shape_binary_broadcast(shapes: Tuple[Shape, ...], attrs: Mapping) -> Shape:
    return broadcast_shapes(shapes[0], shapes[1])


register("add", 2, c_name="bafe_add")(_shape_binary_broadcast)
register("mul", 2, c_name="bafe_mul")(_shape_binary_broadcast)
register("sub", 2, c_name="bafe_sub")(_shape_binary_broadcast)


@register("scale", 2, c_name="bafe_scale")
def _shape_scale(shapes, attrs):
    t, s = shapes
    if not s.is_scalar:
        raise ValueError(f"scale second arg must be scalar, got {s}")
    return t


@register("bias_add", 2, c_name="bafe_bias_add")
def _shape_bias_add(shapes, attrs):
    t, b = shapes
    if t.rank == 0:
        raise ValueError("bias_add needs rank>=1 input")
    if b.rank != 1:
        raise ValueError(f"bias must be rank-1, got {b}")
    if t.dims[-1] != b.dims[0]:
        raise ValueError(f"bias_add last-dim mismatch: {t} vs {b}")
    return t


def _shape_unary_passthrough(shapes, attrs):
    return shapes[0]


register("relu", 1, c_name="bafe_relu")(_shape_unary_passthrough)
register("sigmoid", 1, c_name="bafe_sigmoid")(_shape_unary_passthrough)
register("tanh", 1, c_name="bafe_tanh")(_shape_unary_passthrough)
register("neg", 1, c_name="bafe_neg")(_shape_unary_passthrough)


@register("transpose", 1, attrs_schema=("perm",), c_name="bafe_transpose")
def _shape_transpose(shapes, attrs):
    s = shapes[0]
    perm = attrs["perm"]
    if len(perm) != s.rank:
        raise ValueError(f"transpose perm length {len(perm)} != rank {s.rank}")
    if sorted(perm) != list(range(s.rank)):
        raise ValueError(f"transpose perm {perm} not a permutation of range({s.rank})")
    return Shape(dims=tuple(s.dims[i] for i in perm))


def _shape_reduce(shapes, attrs):
    s = shapes[0]
    axes = attrs.get("axes", tuple(range(s.rank)))
    keepdims = attrs.get("keepdims", False)
    return reduce_shape(s, axes, keepdims=keepdims)


register("reduce_sum", 1, attrs_schema=("axes", "keepdims"), c_name="bafe_reduce_sum")(_shape_reduce)
register("reduce_max", 1, attrs_schema=("axes", "keepdims"), c_name="bafe_reduce_max")(_shape_reduce)


@register("reshape", 1, attrs_schema=("shape",), c_name="bafe_reshape")
def _shape_reshape(shapes, attrs):
    in_shape = shapes[0]
    target = Shape(dims=tuple(attrs["shape"]))
    in_numel = in_shape.numel
    if -1 in target.dims:
        known = 1
        for d in target.dims:
            if d != -1:
                known *= d
        if known == 0 or in_numel % known != 0:
            raise ValueError(f"reshape {in_shape} -> {target} with -1 invalid")
        inferred = in_numel // known
        target = Shape(dims=tuple(inferred if d == -1 else d for d in target.dims))
    if target.numel != in_numel:
        raise ValueError(f"reshape numel mismatch: {in_shape} ({in_numel}) -> {target} ({target.numel})")
    return target


@register("broadcast_to", 1, attrs_schema=("shape",), c_name="bafe_broadcast_to")
def _shape_broadcast_to(shapes, attrs):
    src = shapes[0]
    target = Shape(dims=tuple(attrs["shape"]))
    if src.rank > target.rank:
        raise ValueError(f"broadcast_to: src rank {src.rank} > target {target.rank}")
    pad = (1,) * (target.rank - src.rank) + src.dims
    for s, t in zip(pad, target.dims):
        if s != 1 and s != t:
            raise ValueError(f"broadcast_to: {src} cannot broadcast to {target}")
    return target


# ---------------------------------------------------------------------------
# Fused ops (synthesized by the rewrite engine)
# ---------------------------------------------------------------------------

register("fused_matmul_relu", 2, has_fusion_form=True, c_name="bafe_fused_matmul_relu")(_shape_matmul)


@register("fused_matmul_bias", 3, has_fusion_form=True, c_name="bafe_fused_matmul_bias")
def _shape_fused_matmul_bias(shapes, attrs):
    mm = _shape_matmul((shapes[0], shapes[1]), attrs)
    b = shapes[2]
    if b.rank != 1 or b.dims[0] != mm.dims[-1]:
        raise ValueError(f"fused_matmul_bias bias mismatch: mm={mm}, bias={b}")
    return mm


register("fused_matmul_bias_relu", 3, has_fusion_form=True, c_name="bafe_fused_matmul_bias_relu")(_shape_fused_matmul_bias)


@register("fused_bias_relu", 2, has_fusion_form=True, c_name="bafe_fused_bias_relu")
def _shape_fused_bias_relu(shapes, attrs):
    return _shape_bias_add(shapes, attrs)


DEFAULT_DTYPE = Dtype.F32
